### PR TITLE
WebAuthn CredentialID field needs to be increased in size

### LINKED ISSUE #20457: Webauthn Credential ID length too short
### Description

Hi all :wave: I recently received a new FIDO2 security key, a SoloKey v2. 

Upon registering it to my Gitea instance, a server error was encountered:
> WebauthnRegisterPost() [E] [62db09ae] CreateCredential: pq: value too long for type character varying(410)

While debugging this, I noticed that a 270 byte credential ID was generated for my instance. It seems that the current maximum credential ID length is 255 bytes:
https://github.com/go-gitea/gitea/blob/886b1de94914186a502ed68a6a281a15f1d8d9a3/models/migrations/v210.go#L28

After some tentative searching through the linked WebAuthn specification, I noticed the following:
> **[Credential ID](https://w3c.github.io/webauthn/#credential-id)**
>
> A probabilistically-unique byte sequence identifying a public key credential source and its authentication assertions. At most 1023 bytes long.

I think this issue could be resolved by changing the maximum credential ID length to 1023 bytes raw, or 1640 bytes base32-encoded (per `head -c 1023 /dev/random | base32 --wrap=0 | wc -c`).

### Gitea Version

1.17.0-rc2

### Can you reproduce the bug on the Gitea demo site?

No - its credential ID was exactly 255 bytes, which probably shouldn't be relied on

### Log Gist

https://gist.github.com/spotlightishere/8bcfd9591ca7ec421873091c77644e91

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

Debian 11

### How are you running Gitea?

Downloaded from https://dl.gitea.io/, running underneath systemd

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by zeripath**
WebAuthn have updated their specification to set the maximum size of the
CredentialID to 1023 bytes. This is somewhat larger than our current
size and therefore we need to migrate.

The PR changes the struct to add CredentialIDBytes and migrates the CredentialID string 
to the bytes field before another migration drops the old CredentialID field. Another migration
renames this field back.

Fix #20457

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
Yes this has a migration. I suggest we just port the back the unnecessary migration that is between these two and/or the associated PR.

**Review by jolheiser — APPROVED** [review]
LGTM, until next time 🤠

**Review by Gusted — APPROVED** [review]
@jolheiser You're just asking for problems 🙊

File: models/migrations/v221.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		_, err := x.Exec("ALTER TABLE webauthn_credential MODIFY credential_id VARCHAR(1640)")
+		if err != nil {
+			return err
+		}
+	case schemas.MSSQL:
+		// This column has an index on it. I could write all of the code to attempt to change the index OR
+		// I could just use recreate table.
+		sess := x.NewSession()
+		if err := sess.Begin(); err != nil {
+			_ = sess.Close()
```
Comment (6543) [inline models/migrations/v221.go]: why not use a defer for sess close?

Comment (zeripath) [reply models/migrations/v221.go]: TBH I just copied this code from v210.go and dropped the second part of that migration... and I didn't think more about it as I knew that this part of code has been proven to work in real-world use. This migration is going back to 1.17 and we need to release that quickly I therefore wanted to use as proven a set of a code as possible. 

I mean I guess we could use defer sess.Close() here. 

I guess we could also attempt to extract out a function that does the column resizing and do it properly even for MSSQL but we'll definitely need to write a test if we do that.

Thinking on a valid test for this migration would be - prepare the table to have one or two entries with the old size, ensure the migration doesn't die, and then just need to test insertion of a huge credential ID at the end.

Comment (zeripath) [reply models/migrations/v221.go]: probably we'd need two tests the case where the column size is the old size and one where the column size is the new size but I don't think that breaks.

**Comment by wxiaoguang** [comment]
LGTM, how about setting the length to a large-enough value, like 8000? There should be no difference between 1640 and 8000 on database side, then no worry about overflowing anymore.

File: models/migrations/v221.go (hunk 1 above)
Comment (6543) [reply models/migrations/v221.go]: @zeripath should I push the defer patch?

Comment (zeripath) [reply models/migrations/v221.go]: If you want

**Comment by zeripath** [comment]
The field is indexed and is searched directly so it's potentially bad on db systems like MySQL which store the value directly within the index. I'm not sure if these would store the max string size eg. 8000 bytes or whether they'd just store the string. If it's just the string then it's no worse but I don't know.

The spec has written clearly that this is at most 1023 bytes - previously it wasn't very clear on this hence the change. I guess the question is are they likely to increase it? My suspicion is we'd get more forewarning in that case.

I think if they do increase it again we're probably gonna need to think again about how we implement this. It could be that at that point we'd need to be doing shasums of the credential id searching for that and then doing equality of the credential id - if only to keep the index from being too large.

**Comment by zeripath** [comment]
And in fact I think we've hit the problem I was referring to.

Looking at the tests I see:

```
       testlogger.go:78: 2022/07/29 14:18:52 ...ations/migrations.go:497:Migrate() [I] Migration[210]: v208 was completely broken - remigrate
        migration_test.go:299: 
            	Error Trace:	migration_test.go:299
            	            				migration_test.go:336
            	Error:      	Received unexpected error:
            	            	migrate: migration[210]: v208 was completely broken - remigrate failed: Error 1071: Specified key was too long; max key length is 3072 bytes
            	Test:       	TestMigrations/Migrate-mysql-1.6.4
```

Sigh I think we're going to need to do a bit more work here.

**Comment by zeripath** [comment]
OK I've had to do something a lot more horrible because of MySQL's index issue.

File: models/auth/webauthn.go [hunk 2]
```diff
@@ -20,14 +20,14 @@ import (
...
+	CredentialID []byte
 }
 
 func (err ErrWebAuthnCredentialNotExist) Error() string {
-	if err.CredentialID == "" {
+	if len(err.CredentialID) == 0 {
 		return fmt.Sprintf("WebAuthn credential does not exist [id: %d]", err.ID)
 	}
-	return fmt.Sprintf("WebAuthn credential does not exist [credential_id: %s]", err.CredentialID)
+	return fmt.Sprintf("WebAuthn credential does not exist [credential_id: %s]", base32.HexEncoding.EncodeToString(err.CredentialID))
```
Comment (Gusted) [inline models/auth/webauthn.go]:
```suggestion
	return fmt.Sprintf("WebAuthn credential does not exist [credential_id_bytes: %s]", base32.HexEncoding.EncodeToString(err.CredentialID))
```

**Review by Gusted — COMMENTED** [review]
Small nit, otherwise L-gtm and verified migration on local instance.

File: models/migrations/v223.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	"xorm.io/xorm"
+)
+
+func renameCredentialIDBytes(x *xorm.Engine) error {
+	// This migration maybe rerun so that we should check if it has been run
+	credentialIDExist, err := x.Dialect().IsColumnExist(x.DB(), context.Background(), "webauthn_credential", "credential_id")
+	if err != nil {
+		return err
+	}
+	if !credentialIDExist {
```
Comment (zeripath) [inline models/migrations/v223.go]:
```suggestion
	if credentialIDExist {
```

File: models/auth/webauthn.go (hunk 2 above)
Comment (zeripath) [reply models/auth/webauthn.go]: So now it's back to credential_id rather than credential_id_bytes...

File: models/migrations/v221_test.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		CredentialIDBytes []byte `xorm:"VARBINARY(1024)"` // CredentialID is at most 1023 bytes as per spec released 20 July 2022
+	}
+
+	// Prepare and load the testing database
+	x, deferable := prepareTestEnv(t, 0, new(WebauthnCredential), new(ExpectedWebauthnCredential))
+	if x == nil || t.Failed() {
+		defer deferable()
+		return
+	}
+	defer deferable()
```
Comment (6543) [inline models/migrations/v221_test.go]:
```suggestion
	x, deferable := prepareTestEnv(t, 0, new(WebauthnCredential), new(ExpectedWebauthnCredential))
	defer deferable()
	if x == nil || t.Failed() {
		return
	}
```

File: models/migrations/v223.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		case setting.Database.UseMySQL:
+			if _, err := sess.Exec("ALTER TABLE `webauthn_credential` CHANGE credential_id_bytes credential_id VARBINARY(1024)"); err != nil {
+				return err
+			}
+		case setting.Database.UseMSSQL:
+			if _, err := sess.Exec("sp_rename 'webauthn_credential.credential_id_bytes', 'credential_id', 'COLUMN'"); err != nil {
+				return err
+			}
+		default:
+			if _, err := sess.Exec("ALTER TABLE `webauthn_credential` RENAME COLUMN credential_id_bytes TO credential_id"); err != nil {
```
Comment (6543) [inline models/migrations/v223.go:77]: does sqlite support rename now ?!?

Comment (zeripath) [reply models/migrations/v223.go:77]: https://www.sqlite.org/lang_altertable.html

Comment (zeripath) [reply models/migrations/v223.go:77]:
```
3. ALTER TABLE RENAME COLUMN
The RENAME COLUMN TO syntax changes the column-name of table table-name into new-column-name. The column name is changed both within the table definition itself and also within all indexes, triggers, and views that reference the column. If the column name change would result in a semantic ambiguity in a trigger or view, then the RENAME COLUMN fails with an error and no changes are applied.

```

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Pull request merged**